### PR TITLE
Validate that tests don't leaks tracers, scopes and pooled objects

### PR DESCRIPTION
**Pull request by felixbarny**
## What does this PR do?

Depends on #1271 

With #1270 we have a good foundation to make sure tracers don't leak to other tests classes. This makes `GlobalTracer` more restrictive so that a tracer can only be set via `init` that throws if the current tracer is not a noop.

I have also added the recycling tests on the `@AfterEach` method of `AbstractInstrumentationTest` which required some adjustments to tests to make sure all transactions and spans are ended and thus recycled after being reported. It also checks the bookkeeper object pool to make sure all objects have returned. This also implicitly tests for scope leaks as those would prevent objects from being recycled.

As a bonus, I've found and fixed a scope leak in the currently unreleased instrumentation of `Executor#executeAll`.

## Checklist

- [ ] This is an enhancement of existing features, or a new feature in existing plugins
  - [ ] I have updated [CHANGELOG.asciidoc](CHANGELOG.asciidoc)
  - [ ] I have added tests that prove my fix is effective or that my feature works
  - [ ] Added an API method or config option? Document in which version this will be introduced
  - [ ] I have made corresponding changes to the documentation
- [ ] This is a bugfix
  - [ ] I have updated [CHANGELOG.asciidoc](CHANGELOG.asciidoc)
  - [ ] I have added tests that would fail without this fix
- [ ] This is a new plugin
  - [ ] I have updated [CHANGELOG.asciidoc](CHANGELOG.asciidoc)
  - [ ] My code follows the [style guidelines of this project](https://github.com/elastic/apm-agent-java/blob/master/CONTRIBUTING.md#java-language-formatting-guidelines)
  - [ ] I have made corresponding changes to the documentation
  - [ ] I have added tests that prove my fix is effective or that my feature works
  - [ ] New and existing [**unit** tests](https://github.com/elastic/apm-agent-java/blob/master/CONTRIBUTING.md#testing) pass locally with my changes
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)
  - [ ] I have updated [supported-technologies.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/docs/supported-technologies.asciidoc)
  - [ ] Added an API method or config option? Document in which version this will be introduced
  - [ ] Added an instrumentation plugin? Describe how you made sure that old, non-supported versions are not instrumented by accident.
- [x] This is something else: Test improvements
    - [ ] ~I have updated [CHANGELOG.asciidoc](CHANGELOG.asciidoc)~

### CODE REVIEW

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/context/Destination.java [hunk 1]
```diff
@@ -143,7 +143,7 @@ public boolean hasContent() {
     @Override
     public void resetState() {
         address.setLength(0);
-        port = -1;
+        port = 0;
```
Comment (eyalkoren) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/impl/context/Destination.java:146]: why?

File: apm-agent-plugins/apm-scheduled-annotation-plugin/src/test/java/co/elastic/apm/agent/spring/scheduled/TimerTaskInstrumentationTest.java [hunk 2]
```diff
@@ -56,14 +55,13 @@ void testTimerTask_scheduleWithFixedRate() throws InterruptedException {
...
     void testTimerTask_scheduleWithFixedDelay() throws InterruptedException {
-        reporter.reset();
         TestTimerTask timerTask = new TestTimerTask();
         Timer timer = new Timer("Timer");
         timer.schedule(timerTask, 1L, 10L);
 
         reporter.awaitUntilAsserted(1000L, () -> {
-            timer.cancel();
             assertThat(reporter.getTransactions()).isNotEmpty();
+            timer.cancel();
```
Comment (eyalkoren) [inline apm-agent-plugins/apm-scheduled-annotation-plugin/src/test/java/co/elastic/apm/agent/spring/scheduled/TimerTaskInstrumentationTest.java:64]: Test is failing. Maybe revert to happen before the assertion and wait 20ms before asserting.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/context/Destination.java (hunk 1 above)
Comment (felixbarny) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/context/Destination.java:146]: It made a test fail. In general, when calling `resetState`, the instance should be in the same state as if it had just been instantiated. The field is initialized with the default value of `0`.

File: apm-agent-plugins/apm-scheduled-annotation-plugin/src/test/java/co/elastic/apm/agent/spring/scheduled/TimerTaskInstrumentationTest.java (hunk 2 above)
Comment (felixbarny) [reply apm-agent-plugins/apm-scheduled-annotation-plugin/src/test/java/co/elastic/apm/agent/spring/scheduled/TimerTaskInstrumentationTest.java]: I rewrote the test a little. It should be much more deterministic and resilient now.

**Pull request merged**